packages/store/src/Buffer.sol

```diff
@@ -0,0 +1,368 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+// Acknowledgements:
+// Based on @dk1a's Slice.sol library (https://github.com/dk1a/solidity-stringutils/blob/main/src/Slice.sol)
+// Inspired by @ethier's DynamicBuffer.sol library (https://github.com/divergencetech/ethier/blob/main/contracts/utils/DynamicBuffer.sol)
+
+// Optimized for allocating a fixed size buffer and writing to it / reading from it.
+
+// First 16 bytes are the pointer to the data, followed by 16 bytes of data length.
+type Buffer is uint256;
```

Comment:
I wonder if 10/10/10 split for ptr/length/capacity would be better, then slicing can be done in-place without copying bytes (since you don't need a new length field in-memory)
[LinkedListLib](https://github.com/brockelmore/memmove/blob/master/src/LinkedList.sol) has a 3-way bitpacking example, although it's not very relevant

Reply:
But also, could you get rid of capacity entirely? it seems like an intermediate internal concept, you really need it only for checking overflow. If optimization converges on appendUnchecked anyways, capacity might be useless?

Reply:
Yeah I agree - the whole point of this library is to make encoding/decoding data in `StoreCore` and the typed libs as cheap as possible, and I think it can be optimized to the use cases we have there instead of trying to be as general as possible. Since in the use cases we have there we always know the capacity in advance and just use the buffer as an intermediate to append to (or slice from), I think we can get rid of the explicit capacity. Mind changing this in a follow up?